keepassxc-browser/content/keepassxc-browser.js

```diff
@@ -1,144 +1,223 @@
+                // Single password field
+                kpxcIcons.addIcon(c.password, kpxcIcons.iconTypes.USERNAME);
+            }
         }
+    };
 
-        if (req.action === 'fill_user_pass_with_specific_login') {
-            kpxc.fillWithSpecificLogin(req.id);
-        } else if (req.action === 'fill_username_password') {
-            _called.manualFillRequested = ManualFill.BOTH;
-            await kpxc.receiveCredentialsIfNecessary();
-            kpxc.fillInFromActiveElement(false);
-        } else if (req.action === 'fill_password') {
-            _called.manualFillRequested = ManualFill.PASS;
-            await kpxc.receiveCredentialsIfNecessary();
-            kpxc.fillInFromActiveElement(false, true); // passOnly to true
-        } else if (req.action === 'fill_totp') {
-            await kpxc.receiveCredentialsIfNecessary();
-            kpxc.fillInFromActiveElementTOTPOnly();
-        } else if (req.action === 'clear_credentials') {
-            kpxcEvents.clearCredentials();
-            return Promise.resolve();
-        } else if (req.action === 'activated_tab') {
-            kpxcEvents.triggerActivatedTab();
-            return Promise.resolve();
-        } else if (req.action === 'ignore_site') {
-            kpxc.ignoreSite(req.args);
-        } else if (req.action === 'check_database_hash' && 'hash' in req) {
-            kpxc.detectDatabaseChange(req);
-        } else if (req.action === 'remember_credentials') {
-            kpxc.contextMenuRememberCredentials();
-        } else if (req.action === 'choose_credential_fields') {
-            kpxcDefine.init();
-        } else if (req.action === 'redetect_fields') {
-            const response = await browser.runtime.sendMessage({
-                action: 'load_settings'
-            });
-            kpxc.settings = response;
-            kpxc.initCredentialFields(true);
-        } else if (req.action === 'show_password_generator') {
-            kpxcPasswordDialog.trigger();
-        } else if (req.action === 'add_username_only_option') {
-            kpxc.addToSitePreferences();
+    const addPasswordIcons = async function(c) {
+        // Show password icons also with forms without any username field
+        if (kpxc.settings.usePasswordGeneratorIcons
+            && ((c.username && c.password) || (!c.username && c.passwordInputs.length > 0))) {
+            for (const input of c.passwordInputs) {
+                kpxcIcons.addIcon(input, kpxcIcons.iconTypes.PASSWORD);
+            }
+        }
+    };
+
+    const addTOTPIcons = async function(c) {
+        if (c.totp && kpxc.settings.showOTPIcon) {
+            kpxcIcons.addIcon(c.totp, kpxcIcons.iconTypes.TOTP);
         }
+    };
+
+    for (const form of kpxcForm.savedForms) {
+        await Promise.all([
+            await addUsernameIcons(form),
+            await addPasswordIcons(form),
+            await addTOTPIcons(form)
```

Comment:
We should probably have some simple functions like `addIcons(form)` and `removeIcons(form)` that handles these triple calls. This would allow for easy changes if we add an icon or something like that.